src/object/obj_rpc.c

```diff
@@ -544,61 +544,147 @@ crt_proc_struct_daos_shard_tgt(crt_proc_t proc, crt_proc_op_t proc_op,
-
 static int
-crt_proc_struct_daos_cpd_sub_head(crt_proc_t proc, crt_proc_op_t proc_op,
-				  struct daos_cpd_sub_head *dcsh, bool mbs)
+crt_proc_struct_dtx_mbs(crt_proc_t proc, crt_proc_op_t proc_op,
+			struct dtx_memberships **p_mbs)
 {
-	uint32_t	size = 0;
-	int		rc;
+	struct dtx_memberships	*mbs = NULL;
+	uint32_t		 size = 0;
+	int			 rc;
 
 	if (FREEING(proc_op)) {
-		if (mbs)
-			D_FREE(dcsh->dcsh_mbs);
+		D_FREE(*p_mbs);
 		return 0;
 	}
 
-	rc = crt_proc_struct_dtx_id(proc, proc_op, &dcsh->dcsh_xid);
+	if (ENCODING(proc_op)) {
+		mbs = *p_mbs;
+		size = sizeof(*mbs) + mbs->dm_data_size;
+	}
+
+	/* Pack the size of mbs to help decode case. */
+	rc = crt_proc_uint32_t(proc, proc_op, &size);
 	if (unlikely(rc))
 		return rc;
 
-	rc = crt_proc_daos_unit_oid_t(proc, proc_op, &dcsh->dcsh_leader_oid);
-	if (unlikely(rc))
+	D_ASSERT(size != 0);
+
+	if (DECODING(proc_op)) {
+		D_ALLOC(mbs, size);
+		if (mbs == NULL)
+			return -DER_NOMEM;
+	}
+
+	rc = crt_proc_memcpy(proc, proc_op, mbs, size);
+	if (unlikely(rc)) {
+		if (DECODING(proc_op))
+			D_FREE(mbs);
 		return rc;
+	}
 
-	rc = crt_proc_memcpy(proc, proc_op,
-			     &dcsh->dcsh_epoch, sizeof(dcsh->dcsh_epoch));
+	if (DECODING(proc_op))
+		*p_mbs = mbs;
+
+	return 0;
+}
+
+static int
+crt_proc_struct_obj_dtx_mbs(crt_proc_t proc, crt_proc_op_t proc_op,
+			    struct obj_dtx_mbs *odm)
+{
+	int	rc;
+
+	rc = crt_proc_struct_dtx_id(proc, proc_op, &odm->odm_xid);
 	if (unlikely(rc))
 		return rc;
 
-	if (!mbs)
+	return crt_proc_struct_dtx_mbs(proc, proc_op, &odm->odm_mbs);
+}
+
+static int
+crt_proc_struct_obj_coll_target(crt_proc_t proc, crt_proc_op_t proc_op, struct obj_coll_target *oct)
+{
+	int	rc;
+
+	if (FREEING(proc_op)) {
+		D_FREE(oct->oct_bitmap);
+		D_FREE(oct->oct_shards);
 		return 0;
+	}
 
-	if (ENCODING(proc_op))
-		/* Pack the size of dcsh->dcsh_mbs to help decode case. */
-		size = sizeof(*dcsh->dcsh_mbs) + dcsh->dcsh_mbs->dm_data_size;
+	rc = crt_proc_uint32_t(proc, proc_op, &oct->oct_rank);
+	if (unlikely(rc))
+		return rc;
 
-	rc = crt_proc_uint32_t(proc, proc_op, &size);
+	rc = crt_proc_uint16_t(proc, proc_op, &oct->oct_bitmap_sz);
 	if (unlikely(rc))
 		return rc;
 
-	D_ASSERT(size != 0);
+	rc = crt_proc_uint16_t(proc, proc_op, &oct->oct_shard_nr);
+	if (unlikely(rc))
+		return rc;
 
 	if (DECODING(proc_op)) {
-		D_ALLOC(dcsh->dcsh_mbs, size);
-		if (dcsh->dcsh_mbs == NULL)
+		D_ALLOC_ARRAY(oct->oct_bitmap, oct->oct_bitmap_sz);
+		if (oct->oct_bitmap == NULL)
 			return -DER_NOMEM;
+
+		D_ALLOC_ARRAY(oct->oct_shards, oct->oct_bitmap_sz << 3);
+		if (oct->oct_shards == NULL)
+			goto out;
 	}
 
-	rc = crt_proc_memcpy(proc, proc_op, dcsh->dcsh_mbs, size);
-	if (unlikely(rc)) {
-		if (DECODING(proc_op))
+	rc = crt_proc_memcpy(proc, proc_op, oct->oct_bitmap,
+			     sizeof(*oct->oct_bitmap) * oct->oct_bitmap_sz);
+	if (unlikely(rc))
+		goto out;
+
+	rc = crt_proc_memcpy(proc, proc_op, oct->oct_shards,
+			     sizeof(*oct->oct_shards) * (oct->oct_bitmap_sz << 3));
```

Comment:
is it possible that only pack valid oct_shards (oct_shard_nr maybe smaller than bitmap_sz << 3), if it need too many trivial codes maybe fine to just pack all.

Reply:
If only pack valid shards' identifiers, then we need to pack related index for each valid shard's identifier, then we can rebuild the sparse shards array when unpacking on the remote peer. That can work, but it is not only more complex, but also consumes more bytes for GX object. So maybe not worth.